File: Hw7 (Java)/src/HW7_171044023/BoardArray1D.java
package HW7_171044023;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;

/**
 * @author Seyda Ozer
 *
 */
public class BoardArray1D extends AbstractBoard{
	
	/**
	 * This array of string is the 1D game board.
	 */
	private String [] board;
	
	/**
	 * This int value is the coordinate of empty cell in the 1d board.
	 */
	private int  empty_cell;
	
	/**
	 * BoardArray1D no-argument constructor
	 */
	public BoardArray1D() {
		super();
	}
	
	/**
	 * BoardArray1D two-arguments constructor
	 * @param newRow int value of row number
	 * @param newColumn int value of column number
	 */
	public BoardArray1D(int newRow, int newColumn) {
		super(newRow, newColumn);	
		board = new String[row*column];
	}
	
	/**
	 * This method returns string representation of BoardArray1D object.
	 *@return string of representation of BoardArray1D object
	 */
	@Override
	public String toString() {
		String res = board[0];
		
		for(int i=0; i<getRow()*getColumn(); i++) {
		
			if(i == 0)	res += " "; 
			else{
				res += board[i];
				res += " ";
			}
			if(i % getColumn() == getColumn()-1)
				res += "\n";
		}
		
		return res;
	}

	/**
	 * This method reads the board from the file given as parameter.
	 * @param file_name string of the file name
	 */
	@Override
	public void readFromFile(String file_name) {
		findSize(file_name);
		
		try {
		    FileReader read = new FileReader(file_name);
	        BufferedReader buffer = new BufferedReader(read);

	        String line;
	       
	        int empty = 0, begin = 0;
	        int i = 0;
	        
	        while ((line = buffer.readLine()) != null) {
	        	for(int j=0; j<getColumn(); j++) {
		        	// find first empty
		        	if(j == 0)	empty = line.indexOf(" ");
		        	
		        	//board == line
		        	if(empty == -1)	board[i] = line.substring(begin, line.length());
		        	else	board[i] = line.substring(begin, empty);

		        	begin = empty+1;
			        //find next empty
		        	empty = line.indexOf(" ", empty+1);
		        	
		        	i++;
	        	}
            }
			buffer.close();
			findEmptyCell();
			calculateWallNumber();
			fillCheckBoard();
		}
		catch (IOException e) {
            System.err.format("IOException: %s%n", e);
        }
	}

	/**
	 * This method writes the board to the file given as parameter.
	 * @param file_name string of the file name
	 */
	@Override
	public void writeToFile(String file_name) {
		try {
			FileWriter write = new FileWriter(file_name);
			BufferedWriter buffer = new BufferedWriter(write);
			
			for(int i=0; i<getRow()*getColumn(); i++) {
				buffer.write(board[i]);
				
				if(i % getColumn() != getColumn()-1 )
					buffer.write(" ");
				
				else if(i % getColumn() == getColumn()-1 )
					buffer.newLine();
			}

			buffer.close();
		}
		catch (IOException e) {
            System.err.format("IOException: %s%n", e);
        }
		
	}

	/**
	 *This method resets the board to the solution.
	 */
	@Override
	public void reset() {	
		int k = 0;
		
		for(int i=0; i<getRow() * getColumn(); i++){ // I filled the board with the numbers in the array(numbers).
				if(board[i].compareTo("00") != 0){
					board[i] = checkBoard[k];
					k++;
				}
		}
		setNumberOfMoves(0);
		
	}

	/**
	 * This method sets the board size to given values.
	 * The values are given as parameters and there are no restrictions on the board size.
	 * The board is reset after this operation.
	 * @param size1 int value of board's first size
	 * @param size2 int value of board's second size
	 */
	@Override
	public void setSize(int size1, int size2) {
		setRow(size1);
		setColumn(size2);
		calculateWallNumber();
		fillCheckBoard();
		reset();
		
	}

	/**
	 * This method makes a move according to the given char parameter.
	 * @param choice int value of user's choice for the move
	 */
	@Override
	public void move(char choice) {
		boolean flag = false;
		
		findEmptyCell();
		switch(choice){
			case 'L':	case 'l':	//LEFT
				if(getEmptyY() > 0)
					flag = movement(empty_cell-1);
				else	System.out.println("LEFT is invalid move!");
			break;

			case 'R':	case 'r':	//RIGHT
				if(getEmptyY() < getColumn()-1)
					flag = movement(empty_cell+1);
				else	System.out.println("RIGHT is invalid move!");
			break;

			case 'U':	case 'u':	//UP
				if(getEmptyX() > 0)
					flag = movement(empty_cell - getColumn());
				else	System.out.println("UP is invalid move!");
			break;

			case 'D':	case 'd':	//DOWN
				if(getEmptyX() < getRow()-1)
					flag = movement(empty_cell + getColumn());
				else	System.out.println("DOWN is invalid move!");
			break;
		}
		if(flag == true){
			findEmptyCell();
			setNumberOfMoves( numberOfMoves() + 1 );
			setLastMove(choice);
		}
		
	}

	/**
	 * This method checks if the board is solved..
	 * @return returns true if the board is a solution
	 */
	@Override
	public boolean isSolved() {
		int k = 0;
		System.out.println();
		for(int i=0; i<getRow() * getColumn(); i++){
				if(board[i].compareTo("00") != 0){
					if( board[i].compareTo(checkBoard[k]) != 0)	return false;
					k++;
				}
			
		}
		return true;
	}

	/**
	 * This method takes two indexes and returns the corresponding cell content.
	 * @param x int value of first index
	 * @param y int value of second index
	 * @return returns the value in the given coordinates.
	 */
	@Override
	public String cell(int x, int y) {
		int index = x * getColumn() + y;
		return board[index];
	}

	/**
	 * This method controls equality.
	 * Two boards are equal, if the boards are the same.
	 * This method does not consider last move or the number of steps.
	 * @param other the object of AbstractBoard
	 * @return returns true if two boards are equal.
	 */
	@Override
	public boolean Equals(AbstractBoard other) {
		
		if(getRow() == other.getRow() && getColumn() ==  other.getColumn()) {
			for(int i=0; i<getRow(); i++) {
				for(int j=0; j<getColumn(); j++) {
					if( this.cell(i, j).compareTo( other.cell(i, j) ) != 0)	return false;
				}
			}
		}
		else	return false;
		
		return true;
	}
	
	/**
	 * This method swaps the empty cell and the cell in the index given as parameter.
	 * @param index int value of the index of the cell to move
	 * @return returning true, if the cell can move.
	 */
	private boolean movement(int index){
		String temp;

		findEmptyCell();
		if(board[index].compareTo("00") != 0){
			temp = board[empty_cell];
			board[empty_cell] = board[index];
			board[index] = temp;
			return true;
		}
		else	System.out.println("Empty cell cannot move!");
		return false;
	}
	
	/**
	 * This method calculates the number of wall in the board.
	 */
	private void calculateWallNumber() {
		int count = 0;
		
		for(int i=0; i<getRow() * getColumn(); i++) {
				if(board[i].compareTo("00") == 0) count++;
		}
		
		setWallNumber(count);
	}
	
	/**
	 * This method finds the empty cell's current coordinates.
	 */
	private void findEmptyCell() {
		int x = 0, y = 0;
			
		for(int i=0; i<getRow()*getColumn(); i++) {
			if(board[i].compareTo("bb") == 0) {
				x = i/column;
				y = i%column;
				empty_cell = i;
			}
		}	
		setEmptyCell(x, y);
	}
	
	/**
	 * This method finds the sizes of the board read from the file.
	 * @param file_name the string of the name of the file to be read
	 */
	private void findSize(String file_name) {
		try {
		    FileReader read = new FileReader(file_name);
	        BufferedReader buffer = new BufferedReader(read);
	        
	        int size1 = 0;
	        int size2 = 1;
	        
	        String line;
	        
	        while ((line = buffer.readLine()) != null) {
	        	if(size1 == 0){
	        		int i = 0;
		        	while(i < line.length() ) {
		        		if(line.charAt(i) == ' ') size2++;
		        		i++;
		        	}
	        	}
	        	size1++;
            }
			buffer.close();
			
			setRow(size1);
			setColumn(size2);
			
			board = new String[getRow() * getColumn()];	
		}
		catch (IOException e) {
            System.err.format("IOException: %s%n", e);
        }
	}
}

File: Hw5 (C++)/boardarray1d.cpp
#include<iostream>
#include <string>
#include<fstream>
#include<cstdlib>
#include<time.h>
using namespace std;
#include "boardarray1d.h"

namespace Boards{
	BoardArray1D::BoardArray1D() : AbstractBoard(){
		finalBoard = new string[0];
		board = new string[0];
	}

	BoardArray1D::BoardArray1D(const BoardArray1D& other) : AbstractBoard(other){
	
		finalBoard = new string[row*column-wall_number];
		for(int i=0; i<row*column-wall_number; i++)
			finalBoard[i] = other.finalBoard[i];

		board = new string[row*column];

		for(int i=0; i<row*column; i++)
			board[i] = other.board[i];
	}

	void BoardArray1D::operator=(const BoardArray1D& other){
		empty_x = other.empty_x;
		empty_y = other.empty_y;
		empty = empty;

		if(row != other.row && column != other.column){
			delete [] finalBoard;
			finalBoard = new string[other.row*other.column-other.wall_number];

			delete [] board;
			board = new string[other.row*other.column];
		}
		
		row = other.row;
		column = other.column;
		wall_number = other.wall_number;

		for(int i=0; i<row*column-wall_number; i++)
			finalBoard[i] = other.finalBoard[i];	

		for(int i=0; i<row*column; i++)
			board[i] = other.board[i];
		
	}

	bool BoardArray1D::operator==(AbstractBoard& other){
		if(this->row == other.getRow() && this->column == other.getColumn()){
			for(int i=0; i<row; i++){
				for(int j=0; j<column; j++){
					if( this->operator()(i, j) != other(i, j) ) return false;
				}
			}
		}
		else	return false;
		return true;
	}

	void BoardArray1D::move(char choice){
		bool flag;
		findEmptyCell();
		switch(choice){
			case 'L':	case 'l':	//LEFT
				if(empty_y > 0)
					flag = movement(empty-1);
				else	cout << "LEFT is invalid move!\n";
			break;

			case 'R':	case 'r':	//RIGHT
				if(empty_y < column-1)
					flag = movement(empty+1);
				else	cout << "RIGHT is invalid move!\n";
			break;

			case 'U':	case 'u':	//UP
				if(empty_x > 0)
					flag = movement(empty-column);
				else	cout << "UP is invalid move!\n";
			break;

			case 'D':	case 'd':	//DOWN
				if(empty_x < row-1)
					flag = movement(empty+column);
				else	cout << "DOWN is invalid move!\n";
			break;
		}
		if(flag == true){
			if(choice == 'L' || choice == 'l')	cout << "your move is Left" << endl;
			else if(choice == 'R' || choice == 'r')	cout << "your move is Right" << endl;
			else if(choice == 'U' || choice == 'u')	cout << "your move is Up" << endl;
			else if(choice == 'D' || choice == 'd')	cout << "your move is Down" << endl;
			number_of_moves ++;
			setLastMove(choice);
		}
		else	cout << "this is invalid move!\n";
	}

	bool BoardArray1D::movement(int i){ // this function swaps the two coordinates.
		string temp;

		findEmptyCell(); // this function find the empty cell's current coordinates.
		if(board[i] != "00"){ // if the block is wall, the empty cell do not move.
			temp = board[empty];
			board[empty] = board[i];
			board[i] = temp;
			return true;
		}
		return false;
	}

	void BoardArray1D::setSize(int size1, int size2){
		row = size1;
		column = size2;
		board = new string[row*column];
		setFinalBoard();
	}

	string BoardArray1D::operator()(const int x, const int y) const{
		int index = x*column + y;
		return board[index];
	}

	bool BoardArray1D::isSolved(){
		//this function checks if this board is solved.
		int num1 = 50, num2 = 49;
		string number = "01";

		int k = 0;
		for(int i=0; i<row*column; i++){
			if(board[i] != "00"){
				if( board[i] !=  finalBoard[k] )	return false;
				k++;
			}
		}
		return true;
	}

	void BoardArray1D::setFinalBoard(){ //this function fills the array accordig to the final board.
		calculateWallNumber();
		finalBoard = new string[row*column-wall_number];

		if(row*column-wall_number > 0){
			int i, j, num1 = 50, num2 = 49;
			string number = "01";

			for(i=0; i<row*column-1-wall_number; i++){ // I filled the array with the numbers in the game with using ascii table. I used the array to compare the correct board with the current board and fill the final board. 
				finalBoard[i] = number;
				if(num1 > 57){
					num1 = 48;
					number[0] = num2;
					num2++;
				}
				number[1] = num1;
				num1++;
			}

			finalBoard[row*column-1-wall_number] = "bb"; // I put 0 on the last element of the array for the empty cell.
		}
	}

	void BoardArray1D::calculateWallNumber(){ // this function calculates the number of wall in the game.
		wall_number = 0;

		for(int i=0; i<row*column; i++)
				if(board[i] == "00") wall_number++;

	}

	void BoardArray1D::print() const{
		int i;
		cout << endl;
		for(i=0; i<row*column; i++){
			cout.width(5);
			if(board[i] == "bb")	cout << " ";
			else if(board[i] == "00")	cout << "*";
			else	cout << board[i];
			if(i % column == column-1)	cout << endl;
		}
		cout << endl;
	}

	void BoardArray1D::readFromFile(string file_name){ // this function reads the file.
		setSize(0, 0);
		string line;
		bool flag;
		int size1 = 0, size2 = 0;

		fstream gameBoard;
		do{
			gameBoard.open(file_name);
			if( gameBoard.is_open() ){
			flag = true;
			}
			else{
				cout << "Please enter the file name(.txt) you want to play." << endl;
				cin >> file_name;
				flag = false;
			}
		}while(flag != true);

		while( getline(gameBoard, line) )
			size1++; // the number of row
		
		int i = 0;
		while(i < 25){
				if(line[i] == ' ')	size2++; // the number of column
				i++;
		}
		size2 += 1;
		gameBoard.close();

		setSize(size1, size2);
		i = 0;
		gameBoard.open(file_name);
		while(gameBoard >> board[i]){
			i++;
		}
		gameBoard.close();

		findEmptyCell();
		setFinalBoard(); /* I call this function again(previously called in setSize(line267))
		because the wall_number was not known because the file was not read.*/
		setNumberOfMoves(0);
	}

	void BoardArray1D::writeToFile() const{
		string file_name;

		cout << "Please enter the file name(.txt) you want to save: " << endl;
		cin >> file_name;
		
		ofstream saveBoard;
		saveBoard.open(file_name);
		for(int i=0; i<row*column; i++){
			if(i % column == column-1)	saveBoard << board[i];
			else	saveBoard << board[i] << " ";
			if(i % column == column-1)	saveBoard << endl;
		}
		saveBoard.close();
	}

	void BoardArray1D::reset(){

		auto k = 0;
		for(int i=0; i<row*column; i++){ // I filled the board with the numbers in the array(numbers).
			if(board[i] != "00"){
				board[i] = finalBoard[k];
				k++;
			}
		}
		setNumberOfMoves(0);
	}

	void BoardArray1D::initRandomBoard(){
		srand(time(NULL));
		int temp[9*9], control_array[row*column], i, j, counter;
		bool flag;
		auto size = 0; // this variable is the size of control_array

		for(i=0; i<row*column; i++){
			do{
				counter = 0;
				temp[i] = rand() % (row*column);
				for(int n=0; n<size; n++)
					if(temp[i] == control_array[n]) counter++; // if the random number exists in the array, this counter increases. so flag is false.
				if(counter == 0)	flag = true;
				else	flag = false;
			}while(flag != true);
			control_array[size] = temp[i];
			size++;
		}

		//fill the board
		string number[row*column];
		int k = 0;
		//I convert integer array to string array.
		for(i=0; i<row*column; i++){
			if(temp[i] > 9){
				number[k] += (temp[i] / 10 + 48);
				number[k] += (temp[i] % 10 + 48);
			}
			else{
				number[k] += '0';
				number[k] += (temp[i] + 48);
			}
			if(number[k] == "00")	board[i] = "bb";
			else	board[i] = number[k];
			k++;
		}

		findEmptyCell();
	}

	void BoardArray1D::findEmptyCell(){ // this function find the empty cell's current coordinates.

		for(int i=0; i<row*column; i++){
			if(board[i] == "bb"){
				empty_x = i/column;
				empty_y = i%column;
				empty = i;
			}
		}
	}

	BoardArray1D::~BoardArray1D(){
		delete [] finalBoard;
		delete [] board;
	}
}